storage/test/merkleTreeAdapter_test.cpp

```diff
@@ -231,43 +606,191 @@ TEST(db_adapter, get_key_by_ver) {
+    const auto status = adapter.getKeyByReadVersion(3, defaultSliver, out, actualVersion);
+    ASSERT_TRUE(status.isOK());
+    ASSERT_TRUE(out == data3);
+    ASSERT_EQ(actualVersion, 3);
   }
 
-  // Get a key past the last one and expect the last one
+  // Get a key past the last one and expect the last one.
   {
     Sliver out;
     BlockId actualVersion;
     const auto status = adapter.getKeyByReadVersion(42, defaultSliver, out, actualVersion);
-    ASSERT_TRUE(status == Status::OK());
-    ASSERT_TRUE(out == data5);
-    ASSERT_EQ(actualVersion, 5);
+    ASSERT_TRUE(status.isOK());
+    ASSERT_TRUE(out == data3);
+    ASSERT_EQ(actualVersion, 3);
   }
 }
 
+TEST_P(db_adapter, add_and_get_block) {
+  for (auto numKeys = 1u; numKeys <= maxNumKeys; ++numKeys) {
+    auto adapter = DBAdapter{GetParam()->db()};
+
+    SetOfKeyValuePairs updates1, updates2;
+    for (auto i = 1u; i <= numKeys; ++i) {
+      // Make the keys overlap between block1 and block2.
+      updates1[getSliverOfSize(i)] = getSliverOfSize(i * 10);
+      updates2[getSliverOfSize(i * 2)] = getSliverOfSize(i * 2 * 10);
+    }
+
+    // Add 2 blocks.
+    {
+      const auto status1 = adapter.addBlock(updates1, 1);
+      ASSERT_TRUE(status1.isOK());
+
+      const auto status2 = adapter.addBlock(updates2, 2);
+      ASSERT_TRUE(status2.isOK());
+    }
+
+    // Try to get a non-existent block.
+    {
+      Sliver block;
+      auto found = true;
+      const auto status = adapter.getBlockById(3, block, found);
+      ASSERT_TRUE(status.isOK());
+      ASSERT_FALSE(found);
+    }
+
+    // Get the first block.
+    Sliver block1;
+    {
+      auto found = false;
+      const auto status = adapter.getBlockById(1, block1, found);
+      ASSERT_TRUE(status.isOK());
+      ASSERT_TRUE(found);
+      ASSERT_TRUE(updates1 == block::getData(block1));
+    }
+
+    // Get the second block.
+    {
+      Sliver block2;
+      auto found = false;
+      const auto status = adapter.getBlockById(2, block2, found);
+      ASSERT_TRUE(status.isOK());
+      ASSERT_TRUE(found);
+      ASSERT_TRUE(updates2 == block::getData(block2));
+      ASSERT_TRUE(adapter.getStateHash() == block::getStateHash(block2));
+      ASSERT_TRUE(block::getParentDigest(block2) == blockDigest(1, block1));
+    }
+  }
+}
+
+TEST_P(db_adapter, add_multiple_deterministic_blocks) {
+  auto adapter = DBAdapter{GetParam()->db()};
+
+  const auto numBlocks = 250;
+  auto count = std::uint16_t{0};
+  for (auto i = 1u; i <= numBlocks; ++i) {
+    ASSERT_TRUE(adapter.addBlock(getDeterministicBlockUpdates(count + 1), i).isOK());
+    ASSERT_EQ(adapter.getLatestBlock(), i);
+    ++count;
+  }
+  ASSERT_EQ(adapter.getLatestBlock(), numBlocks);
+
+  count = 0;
+  for (auto i = 1u; i <= numBlocks; ++i) {
+    auto block = Sliver{};
+    auto found = false;
+    ASSERT_TRUE(adapter.getBlockById(i, block, found).isOK());
+    ASSERT_FALSE(block.empty());
+    ASSERT_TRUE(found);
+
+    if (i > 1) {
```

Comment:
The check is there, because we expect an empty parent digest for block 1, not because of `i - 1`. I've reordered it and added a comment to make it clearer. Please let me know if I've understood it correctly.